samples/torch/src/main/kotlin/HelloTorch.kt

```diff
@@ -0,0 +1,104 @@
+import kotlinx.cinterop.*
+import torch.*
+
+abstract class FloatTensor(val raw: CPointer<THFloatTensor>) {
+    private val storage: CPointer<THFloatStorage> get() = raw.pointed.storage!!
+    private val elements get() = storage.pointed
+    private val data: CPointer<FloatVar> get() = elements.data!!
+    private val size: CPointer<LongVar> get() = raw.pointed.size!!
+    private val nDimension: Int get() = raw.pointed.nDimension
+    val shape: List<Int> = (0 until nDimension).map { size[it].toInt() }
+
+    fun sum() = THFloatTensor_sumall(raw)
+    fun flatten() = (0 until elements.size).map { data[it] }.toTypedArray()
+
+    fun dispose() {
+        THFloatTensor_free(raw)
+    }
+
+    abstract override fun toString(): String
+}
+
+class FloatVector(raw: CPointer<THFloatTensor>) : FloatTensor(raw) {
+    operator fun get(i: Int) = THFloatTensor_get1d(raw, i.toLong())
+    operator fun set(i: Int, value: Float) = THFloatTensor_set1d(raw, i.toLong(), value)
+    fun toArray() = (0 until shape[0]).map { i0 -> this[i0] }.toTypedArray()
+
+    operator fun times(other: FloatVector) = THFloatTensor_dot(raw, other.raw)
+
+    fun abs() = kotlin.math.sqrt(this * this)
+
+    override fun toString() = "[${toArray().joinToString { it.toString() }}]"
+}
+
+class FloatMatrix(raw: CPointer<THFloatTensor>) : FloatTensor(raw) {
+    operator fun get(i0: Int, i1: Int) = THFloatTensor_get2d(raw, i0.toLong(), i1.toLong())
+    operator fun set(i0: Int, i1: Int, value: Float) = THFloatTensor_set2d(raw, i0.toLong(), i1.toLong(), value)
+    fun toArray() = (0 until shape[0]).map { i0 -> (0 until shape[1]).map { i1 -> this[i0, i1] }.toTypedArray() }.toTypedArray()
+
+    operator fun times(vector: FloatVector): FloatVector {
+        val result = THFloatTensor_newWithSize1d(shape[0].toLong())!!
+        val added = zeros(shape[0])
+        THFloatTensor_addmv(result, 1f, added.raw, 1f, raw, vector.raw)
+        added.dispose()
+        return FloatVector(result)
+    }
+
+    operator fun times(matrix: FloatMatrix): FloatMatrix {
+        val result = THFloatTensor_newWithSize2d(shape[0].toLong(), matrix.shape[1].toLong())!!
+        val added = zeros(shape[0], matrix.shape[1])
+        THFloatTensor_addmm(result, 1f, added.raw, 1f, raw, matrix.raw)
+        added.dispose()
+        return FloatMatrix(result)
+    }
+
+    override fun toString() = "[${toArray().joinToString(",\n") { "[${it.joinToString { it.toString() }}]" }}]"
+}
+
+fun tensor(vararg values: Float): FloatVector {
+    val tensor = FloatVector(THFloatTensor_newWithSize1d(values.size.toLong())!!)
```

Comment:
I would suggest to rework example a bit, to make tensor disposal more automated, see for example https://github.com/JetBrains/kotlin-native/blob/master/samples/videoplayer/src/main/kotlin/Disposable.kt and its usage.

Reply:
Also using signExtend() instead of toLong() may make code more portable.